Managers/UInputMan.cpp

```diff
@@ -1,2858 +1,1393 @@
+					if (element[INPUT_R_UP].JoyDirMapped()) {
+						if (joy[joystick].stick[element[INPUT_R_UP].GetStick()].flags & JOYFLAG_UNSIGNED) {
+							joy[joystick].stick[element[INPUT_R_UP].GetStick()].axis[element[INPUT_R_UP].GetAxis()].pos = 128;
+							joy[joystick].stick[element[INPUT_R_UP].GetStick()].axis[element[INPUT_R_UP].GetAxis()].d1 = 0;
+							joy[joystick].stick[element[INPUT_R_UP].GetStick()].axis[element[INPUT_R_UP].GetAxis()].d2 = 0;
+						} else {
+							joy[joystick].stick[element[INPUT_R_UP].GetStick()].axis[element[INPUT_R_UP].GetAxis()].pos = 0;
+							joy[joystick].stick[element[INPUT_R_UP].GetStick()].axis[element[INPUT_R_UP].GetAxis()].d1 = 0;
+							joy[joystick].stick[element[INPUT_R_UP].GetStick()].axis[element[INPUT_R_UP].GetAxis()].d2 = 0;
+						}
+					}
+				}
+			}
+
+			// Stick, axis, directions
+			for (int stick = 0; stick < joy[joystick].num_sticks; ++stick) {
+				for (int axis = 0; axis < joy[joystick].stick[stick].num_axis; ++axis) {
+					// Adjust joystick values to eliminate values in deadzone
+					if (player > -1 && deadZoneType == DeadZoneType::SQUARE && deadZone > 0.0F) {
+						// This one heavily relies on AnalogAxiesValue method of processing joystick data.
+						if (joy[joystick].stick[stick].flags & JOYFLAG_UNSIGNED) {
+							if (std::abs(joy[joystick].stick[stick].axis[axis].pos - 128) > 0 && std::abs(joy[joystick].stick[stick].axis[axis].pos - 128) / 128 < deadZone) {
+								joy[joystick].stick[stick].axis[axis].pos = 128;
+								joy[joystick].stick[stick].axis[axis].d1 = 0;
+								joy[joystick].stick[stick].axis[axis].d2 = 0;
+							}
+						} else {
+							if (std::abs(joy[joystick].stick[stick].axis[axis].pos) > 0 && std::abs(joy[joystick].stick[stick].axis[axis].pos) / 128 < deadZone) {
+								joy[joystick].stick[stick].axis[axis].pos = 0;
+								joy[joystick].stick[stick].axis[axis].d1 = 0;
+								joy[joystick].stick[stick].axis[axis].d2 = 0;
+							}
+						}
+					}
+					s_ChangedJoystickStates[joystick].stick[stick].axis[axis].d1 = joy[joystick].stick[stick].axis[axis].d1 != s_PrevJoystickStates[joystick].stick[stick].axis[axis].d1;
+					s_ChangedJoystickStates[joystick].stick[stick].axis[axis].d2 = joy[joystick].stick[stick].axis[axis].d2 != s_PrevJoystickStates[joystick].stick[stick].axis[axis].d2;
+				}
+			}
+		}
+	}
 
-	return m_aNetworkAccumulatedElementReleased[element];
-}
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
 
+	void UInputMan::StoreInputEventsForNextUpdate() {
+		// Save the current state of the keyboard so that we can compare it next frame and see which key states have been changed in the mean time.
+		std::memcpy(s_PrevKeyStates, const_cast<const char *>(key), KEY_MAX);
 
+		// Save the mouse button states so that we can compare it next frame and see which buttons have changed.
+		for (int mouseButton = 0; mouseButton < MAX_MOUSE_BUTTONS; ++mouseButton) {
```

Comment:
Why would we? What benefit do we get from this over regular assignment?
For the key array i'm guessing memcpy is used because it's volatile and/or other fucky Allegro stuff.

Reply:
I should have said `std::copy` rather than so than `std::memcpy` actually. As for the benefits, potentially a tiny bit faster (which isn't important since we're only iterating through a few things), less code/cleaner looking, etc.
https://stackoverflow.com/questions/29099564/what-is-the-benefit-of-using-stdcopy-instead-of-manual-for-loop-to-copy-dynami
Ultimately not a big deal either way, I mostly saw the memcpy there and wondered why not do it elsewhere too. Don't bother changing it if you don't want to.

Reply:
So it should look like this?
`std::copy(std::begin(s_MouseButtonStates), std::end(s_MouseButtonStates), std::begin(s_PrevMouseButtonStates));`

What is we use `std::swap`?
`std::swap(s_MouseButtonStates, s_PrevMouseButtonStates);`

Since `s_MouseButtonStates` values are being overwritten during update it shouldn't matter that we shoved in the ones from last update.
Thoughts?

Reply:
For copy it's the other way around - `std::copy(first, last, destinationFirst);`. destinationLast here is automatically handled based on what you've specified you're copying.

That said, yeah, if you can use std::swap instead without having any problems then go for that cause that's super smort.